segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java

```diff
@@ -1039,20 +1052,33 @@ private CompletableReadResultEntry createRedirectedRead(long streamSegmentOffset
-            // case, then this entry will fail with either ObjectClosedException or StreamSegmentNotFoundException, since
-            // it is pointing to the now defunct Transaction segment. At that time, a simple retry of the read would
-            // yield the right result. However, in order to recover from this without the caller's intervention, we pass
-            // a pointer to getSingleReadResultEntry to the RedirectedReadResultEntry in case it fails with such an exception;
-            // that class has logic in it to invoke it if needed and get the right entry.
-            result = new RedirectedReadResultEntry(result, entry.getStreamSegmentOffset(),
-                    (rso, ml, sourceSegmentId) -> getOrRegisterRedirectedRead(rso, ml, sourceSegmentId, makeCopy), redirectedIndex.metadata.getId());
-        }
 
-        return result;
+        try {
+            CompletableReadResultEntry result = redirectedIndex.getSingleReadResultEntry(redirectOffset, maxLength, makeCopy);
+            if (result != null) {
+                // Since this is a redirect to a (merged) Transaction, it is possible that between now and when the caller
+                // invokes the requestContent() on the entry the Transaction may be fully merged (in Storage). If that's the
+                // case, then this entry will fail with either ObjectClosedException or StreamSegmentNotFoundException, since
+                // it is pointing to the now defunct Transaction segment. At that time, a simple retry of the read would
+                // yield the right result. However, in order to recover from this without the caller's intervention, we pass
+                // a pointer to getSingleReadResultEntry to the RedirectedReadResultEntry in case it fails with such an exception;
+                // that class has logic in it to invoke it if needed and get the right entry.
+                result = new RedirectedReadResultEntry(result, entry.getStreamSegmentOffset(),
+                        (rso, ml, sourceSegmentId) -> getOrRegisterRedirectedRead(rso, ml, sourceSegmentId, makeCopy), redirectedIndex.metadata.getId());
+            }
+
+            return result;
+        } catch (ObjectClosedException ex) {
+            // Similarly to above, but the source Segment (Transaction) has been merged between when getSingleReadResultEntry
+            // (our caller) and redirectedIndex.getSingleReadResultEntry() was invoked. Since we are not holding the lock
+            // while executing this method (due to deadlock considerations), this scenario is plausible. If this does
+            // indeed happen, we need to reissue the read against ourself, in which case an updated result/entry will be
+            // returned.
+            if (!redirectedIndex.closed) {
+                throw ex;
+            }
+            return getSingleReadResultEntry(streamSegmentOffset, maxLength, makeCopy);
```

Comment:
In line with @fpj comment, it may be the case that this catch block does not does what we expect, despite you have already mentioned why testing this may be hard to do.

Reply:
@RaulGracia @fpj Isn't that comment explanatory enough? I cannot write a unit test to deterministically test this path because of this (quote from the comment): `the source Segment (Transaction) has been merged between when getSingleReadResultEntry (our caller) and redirectedIndex.getSingleReadResultEntry() was invoked`.

Those calls execute on the same thread in short sequence, with no breaks. It was the case that in the past they were both protected by a mutex, but I had to remove that protection because of a deadlock we found a while back. So the second call happens outside of that lock. Without introducing code hooks (empty methods that I override in special, test-only subclasses), I cannot stop the execution of the thread so I can simulate this kind of scenario (i.e., merge the other segment). In the real world, the merger happens async, triggered by the OperationProcessor, which runs separately from read execution

Reply:
I asked a different question:

https://github.com/pravega/pravega/pull/5472#issuecomment-758589501

Reply:
@fpj I added a pretty lengthy comment inside that catch block explaining how we get there. I believe that is pretty self-explanatory.

Reply:
The confusion for me is that there are many issues conflated in one PR, and I initially didn't understand why we had changes in `main` code while the code changes are not being tested. You have edited the description to clarify some of the changes, and you have also clarified offline that you are working on a single branch to fix a series of issues with flaky tests and bugs. Thanks for both clarifications.